### ISSUE TITLE
Feature Request: Adjust popup page layouts

### ISSUE THREAD

**Issue by roderik-maker** [Blackymas/NSPanel_HA_Blueprint#1329]
I am super grateful for this awesome project, thank you to everyone who has contributed. 

A quality of life feature I would like to see implemented would be to make the popups for entities like lights more configurable. The settings I would like to be able to configure are:

- Rotate the sliders by 90degrees. -_This would also make the "adaptive hardware button actions" (from feature request [1014](https://github.com/Blackymas/NSPanel_HA_Blueprint/issues/1014), if it is ever implemented) more intuitive for controlling sliders._ 
- Show 2 sliders on the same screen. - _ie. Show both the light temperature and light brightness sliders on the first popup page._
- Change the default page shown in a popup. - _ie. Show the light RGB page as the default instead of the brightness slider._

**Comment by edwardtfn** [comment]
It would probably be easier if we split this in 3 individual Feature Requests...

But let's go...

> Rotate the sliders by 90degrees. -This would also make the "adaptive hardware button actions" (from feature request https://github.com/Blackymas/NSPanel_HA_Blueprint/issues/1014, if it is ever implemented) more intuitive for controlling sliders.

This might be feasible with the EU model, but it could be an issue with the US version. Having 2 different ways is possible, but it will require having distinct pics (the slider is one pic overlapping another pic) and will increase the number of points for maintenance.
By the way, when I open the light using Home Assistant it shows the slider in the vertical. Shouldn't be better to keep that same behavior?

> Show 2 sliders on the same screen. - ie. Show both the light temperature and light brightness sliders on the first popup page.

This is feasible, but again will be different as it is in Home Assistant. Not that I think it should always be the same, but the more we deviate from what people is used of, the more we will have new Feature Requests asking for changing it. 😉 
Still about this... how should the buttons behave with 2 sliders in case #1014 is implemented?

> Change the default page shown in a popup. - ie. Show the light RGB page as the default instead of the brightness slider.

Probably the easier of these 3 requests, and probably easy to have a way for each user customize which should be shown first.
Again, on Home Assistant, it always open with brightness, when available.

**Comment by roderik-maker** [comment]
Thank you for the extensive response. I'll do my best to be as thorough in my reply. 

**Point 1**
> This might be feasible with the EU model, but it could be an issue with the US version. Having 2 different ways is possible, but it will require having distinct pics (the slider is one pic overlapping another pic) and will increase the number of points for maintenance.

I am not very familiar with UI design, certainly not for these types of devices, but I gather from your comment that there is no equivalent to the HTML 'rotate()' function to display the slider? Does the US_land config not have asset's that might be repurposed for this?
I don't have a US version of the panel but I understand the desire to keep the panels as similar as possible. (I would have thought the US version would be great and might even  fit 3 vertically but I follow your reasoning that this would require the sliders to either be unpractically narrow or edge to edge and ugly.)

> By the way, when I open the light using Home Assistant it shows the slider in the vertical. Shouldn't be better to keep that same behaviour?

On a phone screen I think the up and down motion is the easiest (users can use their thumb). However for a wall display, when bracing your wrist on the wall to interact with the panel, I think a horizontal slider is easier to control as you can leave your wrist in the same place. Maybe I just have to get more used to the panel but sometimes my hand slips when dragging my brightness slider in the morning (which is not particularly pleasant). In the end I suppose it really is just a matter of personal preference and potentially also the panels height on the wall etc... (I am also not suggesting a global change, just giving users the option to pick what works best for them.)

**Point 2**
> This is feasible, but again will be different as it is in Home Assistant. Not that I think it should always be the same, but the more we deviate from what people is used of, the more we will have new Feature Requests asking for changing it. 

I follow your reasoning and matching the panels to HA has a lot of value, I am also a big fan of uniformity in design. However if the screen real estate is available, reducing the number of clicks needed for actions also has value (especially as the NSpanel is not perfect at registering touch interactions). _Actually, having the single slider in HA is also a pet peeve of mine for the exact same reason_. I understand the necessity for a single slider on very small displays (to limit accidental touches) but I think the NS panel is large enough to show both sliders without looking too bussy or resulting in input errors. Both options can make sense and just like the point before, I think empowering the end user to choose for themselves is the best of both worlds. 

But as development time is a very precious resource I'll just hope HA sees the light and adds the second slider to the more-info panel haha (although I doubt it will happen any time soon with their recent card designs veering more towards minimalism). 

Side note: I know my way around python, made a fair share of esphome projects and played around with html, (as well as some other object oriented languages) any idea how long it would take me to dive in and familiarise myself with the code/language you are using to try and see if I can implement my suggested changes myself? (I know its a hard question to answer but a link to some learning resources would be greatly appreciated. No promises though as it seems to be quite a bit outside of the types of code I am familiar with).

> Still about this... how should the buttons behave with 2 sliders in case https://github.com/Blackymas/NSPanel_HA_Blueprint/issues/1014 is implemented?

I also chipped in on that [thread](https://github.com/Blackymas/NSPanel_HA_Blueprint/issues/1014) on how I thought variable button actions could be implemented:
Allow users to defined left and right button actions for each unique page on the panel. If no alternative button actions are provided for a page, the buttons are mapped to the default actions.

This way users can choose for themselves which of the two sliders the buttons control or if they want the buttons to do something else entirely. (You can probably tell by now, I am a big fan of providing options ;) )

**Point 3**
> Probably the easier of these 3 requests, and probably easy to have a way for each user customise which should be shown first.
Again, on Home Assistant, it always open with brightness, when available.

I was hoping that this was indeed just changing the redirect link for long presses. But to be honest I also added it as I thought it would be an easy setting to implement should you decide to give users more freedom to determine the layouts. For me personally this is something I would use for a few light entities but if it was a global parameter I would keep it set to brightness as that is what I use the most for most light entities.

Thank you for all the hard work!!!!
 
_My mother-tongues are dutch and english which are over represented in this community so I don't think I can help with translation and I am unfamiliar with the code/language you are using for the panel configuration. (I tried looking through the .txt files in nspanel_eu_code but found it hard to follow...) Nevertheless, let me know if you think there's something I could do to help with the project._

Ps. Sorry for the essay, I wanted to be thorough, hope I wasn't repetative

**Comment by edwardtfn** [comment]
> there is no equivalent to the HTML 'rotate()' function to display the slider?

In fact, there is a slider component which can be set as horizontal or vertical, but that component, by default, is quite square and ugly, so we use picture overlapping with that slider component. Basically, this means we have a picture which consists of the slider "empty" and another picture with the slider "filled", and Nextion takes care of overlapping those as the user changes the position. Ah, and one more picture for the slider tip, in some cases.
In order to have a different direction for the EU model, we have to create those slider images with the other direction. This is not a big deal, but requires some work to have that looking nice, taking the background in account (as this Nextion model doesn't supports transparency).
It is feasible, it's just a bit of UI work.

**Comment by edwardtfn** [comment]
> On a phone screen I think the up and down motion is the easiest (users can use their thumb). However for a wall display, when bracing your wrist on the wall to interact with the panel, I think a horizontal slider is easier to control as you can leave your wrist in the same place.

I agree with you... for the EU model, the horizontal slider will probably be nicer.

> In the end I suppose it really is just a matter of personal preference and potentially also the panels height on the wall etc... (I am also not suggesting a global change, just giving users the option to pick what works best for them.)

This is another challenge imposed by Nextion. All those components are defined on the construction of the TFT and having a user's selectable component would require designing duplicated controls in all those sliders, which will take more memory (also very limited by Nextion).

**Comment by edwardtfn** [comment]
> Side note: I know my way around python, made a fair share of esphome projects and played around with html, (as well as some other object oriented languages) any idea how long it would take me to dive in and familiarise myself with the code/language you are using to try and see if I can implement my suggested changes myself? (I know its a hard question to answer but a link to some learning resources would be greatly appreciated. No promises though as it seems to be quite a bit outside of the types of code I am familiar with).

Well, we have 3 components here:
- Blueprint: Basically YAML and Jinja2 (which is based on Python).
- ESPHome: It is organized with YAML, but in reality most of the code is C++.
- TFT: It's a proprietary format from Nextion. The big pain here is their editor, which is available only for Windows and is far from the best thing to play with. This is where most of the UI is.

It's hard to me to say how dificult it would be for you. If you wanna play with that, I would love to support... just come with your questions and I will do my best.
It's always good to have more hands on this. 😉

**Comment by edwardtfn** [comment]
> I was hoping that this was indeed just changing the redirect link for long presses. But to be honest I also added it as I thought it would be an easy setting to implement should you decide to give users more freedom to determine the layouts. For me personally this is something I would use for a few light entities but if it was a global parameter I would keep it set to brightness as that is what I use the most for most light entities.

The main issue is the number of inputs on the blueprint. It's a lot already, and as that doesn't supports conditionals, we will need lots of additional inputs for each button (we currently have almost 40 buttons). This will certainly degrade the usability and the performance of that settings page.
I think global settings are viable, but settings per button would be quite complex.

**Comment by edwardtfn** [comment]
> I tried looking through the .txt files in nspanel_eu_code but found it hard to follow...

Don't waste time with those txt files. Those are just an export from Nextion proprietary format to Text in a tentative to have some versioning control on GitHub.
The real code resides on the HMI files, which requires [Nextion Editor](https://nextion.tech/nextion-editor/).

**Comment by edwardtfn** [comment]
> Ps. Sorry for the essay, I wanted to be thorough, hope I wasn't repetative

No need to apologies. It's nice to have this kind of discussion and open the mind for new ideas. 😉 
Keep doing that...
